File: alchemist-implementationbase/src/main/java/it/unibo/alchemist/model/timedistributions/Trigger.java
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.timedistributions;

import it.unibo.alchemist.model.times.DoubleTime;
import it.unibo.alchemist.model.Environment;
import it.unibo.alchemist.model.Node;
import it.unibo.alchemist.model.Time;

/**
 * @param <T>
 *            Concentration type
 */
public final class Trigger<T> extends AbstractDistribution<T> {

    private static final long serialVersionUID = 5207992119302525618L;
    private boolean dryRunDone;

    /**
     * @param event
     *            the time at which the event will happen
     */
    public Trigger(final Time event) {
        super(event);
    }

    @Override
    public double getRate() {
        return Double.NaN;
    }

    @Override
    protected void updateStatus(
        final Time currentTime,
        final boolean executed,
        final double param,
        final Environment<T, ?> environment
    ) {
        if (dryRunDone && currentTime.compareTo(getNextOccurence()) >= 0 && executed) {
            setNextOccurrence(new DoubleTime(Double.POSITIVE_INFINITY));
        }
        dryRunDone = true;
    }

    @Override
    public Trigger<T> cloneOnNewNode(final Node<T> destination, final Time currentTime) {
        return new Trigger<>(getNextOccurence());
    }

}

File: alchemist-implementationbase/src/main/kotlin/it/unibo/alchemist/model/timedistributions/Trigger.kt
/*
 * Copyright (C) 2010-2023, Danilo Pianini and contributors
 * listed, for each module, in the respective subproject's build.gradle.kts file.
 *
 * This file is part of Alchemist, and is distributed under the terms of the
 * GNU General Public License, with a linking exception,
 * as described in the file LICENSE in the Alchemist distribution's top directory.
 */
package it.unibo.alchemist.model.timedistributions

import it.unibo.alchemist.model.Environment
import it.unibo.alchemist.model.Node
import it.unibo.alchemist.model.Time
import it.unibo.alchemist.model.times.DoubleTime

/**
 * A trigger that fires at a given [time].
 */
class Trigger<T>(time: Time) : AbstractDistribution<T>(time) {
    override fun getRate(): Double = Double.Companion.NaN

    override fun updateStatus(currentTime: Time?, executed: Boolean, param: Double, environment: Environment<T, *>) {
        if (executed) {
            setNextOccurrence(DoubleTime(Double.Companion.POSITIVE_INFINITY))
        }
    }

    override fun cloneOnNewNode(destination: Node<T>, currentTime: Time): AbstractDistribution<T> =
        Trigger(nextOccurence)

    /**
     * Static constants for [Trigger] class.
     */
    companion object {
        private const val serialVersionUID = 5207992119302525618L
    }
}
